app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/TooltipWidget.java

```diff
@@ -0,0 +1,120 @@
+package org.mozilla.vrbrowser.ui.widgets;
+
+import android.content.Context;
+import android.graphics.PointF;
+import android.graphics.Rect;
+import android.view.View;
+import android.view.ViewGroup;
+import android.view.ViewTreeObserver;
+import android.widget.TextView;
+
+import org.mozilla.vrbrowser.R;
+import org.mozilla.vrbrowser.utils.ViewUtils;
+
+public class TooltipWidget extends UIWidget {
+
+    private View mTargetView;
+    private UIWidget mParentWidget;
+    protected TextView mText;
+    private PointF mTranslation;
+    private float mRatio;
+    private int mPaddingH;
+    private int mPaddingV;
+
+    public TooltipWidget(Context aContext) {
+        super(aContext);
+
+        initialize();
+    }
+
+    private void initialize() {
+        inflate(getContext(), R.layout.tooltip, this);
+
+        mText = findViewById(R.id.tooltipText);
+
+        ViewGroup layout = findViewById(R.id.layout);
+        mPaddingH = layout.getPaddingStart() + layout.getPaddingEnd();
+        mPaddingV = layout.getPaddingTop() + layout.getPaddingBottom();
+    }
+
+    @Override
+    protected void initializeWidgetPlacement(WidgetPlacement aPlacement) {
+        aPlacement.visible = false;
+        aPlacement.width =  0;
+        aPlacement.height = 0;
+        aPlacement.parentAnchorX = 0.0f;
+        aPlacement.parentAnchorY = 1.0f;
+        aPlacement.anchorX = 0.5f;
+        aPlacement.anchorY = 0.5f;
+        aPlacement.translationZ = WidgetPlacement.unitFromMeters(getContext(), R.dimen.tooltip_z_distance);
+    }
+
+    @Override
+    public void show(@ShowFlags int aShowFlags) {
+        measure(MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED),
+                MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED));
+        mWidgetPlacement.translationX = mTranslation.x * (mRatio / mWidgetPlacement.density);
```

Comment:
Would it be possible to use `painter.getTextBounds()` or `painter.measureText()` here to avoid waiting for `OnGlobalLayoutListener` and doing two texture resizes?

Reply:
I couldn't get reliable measures from getTextBounds/measureText but getMeasuredX seems to return legit values after a measure pass and we don't need any further texture resizes. Also I've improved some calculations to get the exact texture sizes.